crates/topos-p2p/src/runtime/mod.rs

```diff
@@ -28,60 +40,99 @@ pub struct Runtime {
+#[derive(Default)]
+pub(crate) struct StateMachine {
+    /// Indicates if the node has external addresses configured
+    pub(crate) has_external_addresses: bool,
+    /// Indicates if the node is listening on any address
+    pub(crate) is_listening: bool,
+    /// List the boot peers that the node has tried to connect to
+    pub(crate) dialed_bootpeer: HashSet<ConnectionId>,
+    /// Indicates if the node has successfully connected to a boot peer
+    pub(crate) successfully_connect_to_bootpeer: Option<ConnectionId>,
+    /// Track the number of retries to connect to boot peers
+    pub(crate) connected_to_bootpeer_retry_count: usize,
+}
+
 impl Runtime {
     /// Bootstrap the p2p layer runtime with the given configuration.
     /// This method will configure, launch and start queries.
     /// The result of this call is a p2p layer bootstrap but it doesn't mean it is
     /// ready.
-    pub async fn bootstrap(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+    pub async fn bootstrap<S: Stream<Item = Event> + Unpin + Send>(
+        mut self,
+        event_stream: &mut S,
+    ) -> Result<JoinHandle<Result<(), P2PError>>, P2PError> {
         debug!("Added public addresses: {:?}", self.public_addresses);
         for address in &self.public_addresses {
             self.swarm.add_external_address(address.clone());
+            self.state_machine.has_external_addresses = true;
         }
-
         debug!("Starting to listen on {:?}", self.listening_on);
+
         for addr in &self.listening_on {
             if let Err(error) = self.swarm.listen_on(addr.clone()) {
                 error!("Couldn't start listening on {} because of {error:?}", addr);
 
-                return Err(Box::new(error));
+                return Err(P2PError::TransportError(error));
             }
-        }
 
-        if !self.peer_set.is_empty() {
-            debug!(
-                "{} Connected to some peers, start a bootstrap query",
-                self.local_peer_id
-            );
-            self.swarm.behaviour_mut().discovery.bootstrap()?;
+            self.state_machine.is_listening = true;
         }
 
-        while let Some(event) = self.swarm.next().await {
-            self.handle(event).await;
+        let mut handle = spawn(self.run().in_current_span());
 
-            if self.swarm.connected_peers().count() >= self.config.minimum_cluster_size {
-                break;
+        // Wait for first healthy
+        loop {
+            tokio::select! {
+                result = &mut handle => {
+                    match result {
+                        Ok(Ok(_)) => info!("P2P layer has been shutdown"),
+                        Ok(Err(error)) => {
+                            error!("P2P layer has failed with error: {:?}", error);
+
+                            return Err(error);
+                        }
+                        Err(_) => {
+                            error!("P2P layer has failed in an unexpected way.");
```

Comment:
Because it's a `JoinError` from `tokio`, it doesn't bring much information you will endup with something like:

`P2P layer has failed in an unexpected way: JoinError::Panic(Id(X), ...)`

or

`P2P layer has failed in an unexpected way: "task X panicked"`

with the panic message being print before in `stderr`